OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,274 @@
+ * numerical data as a table. A DataTable is independent of the data source 
+ * used to populate it. A concrete DataTable provides fast access to data
+ * elements by row and/or column indices as well as columns by label.
+ *
+ * @author Ajay Seth
+ */
+class AbstractDataTable {
+public:
+    virtual ~AbstractDataTable() = default;
+
+    virtual AbstractDataTable* clone() const = 0;
+
+    /** Read the meta data associated with this DataTable as an XMLDocument */
+    virtual const XMLDocument& getMetaData() const = 0;
+    /** Update the meta data associated with this DataTable as an XMLDocument */
+    virtual XMLDocument& updMetaData() = 0;
+
+    /** Get the number of rows in the DataTable */
+    virtual size_t  getNumRows() const = 0;
+    /** Get the number of columns in the DataTable */
+    virtual size_t  getNumCols() const = 0;
+
+    /** Determine if the DataTable has a column associated with a 
+        column label (string) */
+    virtual bool hasColumn(const std::string& label) const = 0;
+    /** For a known column label return its column index. */
+    virtual size_t getColumnIndex(const std::string& label) const = 0;
+
+    /** Read-only column labels associated with the columns of the DataTable */
+    virtual const Array<std::string>& getColumnLabels() const = 0;
+    /** Update column labels associated with the columns of the DataTable */
+    virtual Array<std::string>& updColumnLabels() = 0;
+
+    /* TODO: Read in the units associated with the columns of the DataTable */
+    /* virtual const Array<Units>& getColumnUnits() const = 0;
+    */
+
+    /** Get the underlying type of data held in the DataTable. Can be any 
+        SimTK numerical type: Real, Vec<M>, Mat<M, N>, SpatialVec, ... */
+    virtual const std::type_index& getDataTypeInfo() const = 0;
+
+    /** Dump the DataTable to an output stream */
+    virtual void dumpToStream(std::ostream &out) const = 0;
+
+//=============================================================================
+};  // END of class AbstractDataTable
+//=============================================================================
+
+
+/** Concrete Implementation of the <tt>DataTable<DataType></tt>
+@tparam DataType  Any numerical type: (SimTK::) Real, Vec<M>, Mat<M,N>, ...*/
+template<typename DataType = SimTK::Real>
+class DataTable_ : public AbstractDataTable {
+//=============================================================================
+// METHODS
+//=============================================================================
+public:
+    virtual ~DataTable_(){};
+
+    /** Construct an empty DataTable */
+    DataTable_() : _type(typeid(DataType)) {}
+
+    /** Construct a pre-sized DataTable to be filled in later. If values
+        are accessed before they are assigned they will be NaN. */
+    DataTable_(size_t nrows, size_t ncols, 
+        const DataType& val = DataType(SimTK::NaN)) :
+        _data(int(nrows), int(ncols), val),
+        _type(typeid(DataType)) {}
```

Comment:
What happens with column labels here? With Storage, if you are building a storage object, you have to populate an Array<string>, then use `Storage::setColumnLabels(myArrayStr)`. Is there a better interface?